ratelimit/redis.go

```diff
@@ -120,78 +117,70 @@ func (c *clusterLimitRedis) startSpan(ctx context.Context, spanName string) func
 //
 // If a context is provided, it uses it for creating an OpenTracing span.
 func (c *clusterLimitRedis) AllowContext(ctx context.Context, clearText string) bool {
-	s := getHashedKey(clearText)
 	c.metrics.IncCounter(redisMetricsPrefix + "total")
+	now := time.Now()
+	finishSpan := c.startSpan(ctx, allowSpanName)
+
+	allow, err := c.allowContext(ctx, clearText)
+
+	finishSpan(err != nil)
+	queryFailure := err != nil
+	c.measureQuery(allowMetricsFormat, allowMetricsFormatWithGroup, &queryFailure, now)
+
+	if err != nil {
+		allow = true // fail open
+	}
+	if allow {
+		c.metrics.IncCounter(redisMetricsPrefix + "allows")
+	} else {
+		c.metrics.IncCounter(redisMetricsPrefix + "forbids")
+	}
+	return allow
+}
+
+func (c *clusterLimitRedis) allowContext(ctx context.Context, clearText string) (bool, error) {
+	s := getHashedKey(clearText)
 	key := c.prefixKey(s)
 
 	now := time.Now()
-	var queryFailure bool
-	defer c.measureQuery(allowMetricsFormat, allowMetricsFormatWithGroup, &queryFailure, now)
-
 	nowNanos := now.UnixNano()
 	clearBefore := now.Add(-c.window).UnixNano()
 
-	count, err := c.allowCheckCard(ctx, key, clearBefore)
+	// drop all elements of the set which occurred before one interval ago.
+	_, err := c.ringClient.ZRemRangeByScore(ctx, key, 0.0, float64(clearBefore))
 	if err != nil {
-		log.Errorf("Failed to get from redis cardinality: %v", err)
-		queryFailure = true
-		// we don't return here, as we still want to record the request with ZAdd, but we mark it as a
-		// failure for the metrics
+		return false, err
+	}
+
+	// get cardinality
+	count, err := c.ringClient.ZCard(ctx, key)
+	if err != nil {
+		return false, err
```

Comment:
most of `return false, err` are changing the behavior of the function from fail to open to fail to closed. As far as I see without impact, because you change the decision to fail open on the caller function, but from reading point of view that can be miss leading, because you read a `allowContect` function with return `allow, error`.

Reply:
It has to return either true or false for the value along with an error. The fail open decision is explicitly made on the caller side and this is non-exported method. I could rewrite it to use return values like
```go
func (c *clusterLimitRedis) allowContext(ctx context.Context, clearText string) (allow bool, err error) {
...
	count, err := c.ringClient.ZCard(ctx, key)
	if err != nil {
		return
	}
...
```
but it does not really change much